EIPS/eip-758.md

```diff
@@ -0,0 +1,99 @@
+      "returnData": "0x000000000000000000000000000000000000000000000000000000000000002a"
+    },
+    "subscription": "0x00000000000000000000000000000b0b"
+  }
+}
+```
+
+The caller receives notifications about their transactions' return data in two cases: first when a transaction is sealed, and again (with an extra `"removed": true` field) if a transaction is affected by a chain reorganization.  Notifications are sent to the client for all transactions submitted from the client that are sealed _after_ subscribing.  As with other subscriptions, the caller can send an `eth_unsubscribe` RPC request to stop receiving push notifications:
+
+```json
+{"jsonrpc": "2.0", "id": 2, "method": "eth_unsubscribe", "params": ["0x00000000000000000000000000000b0b"]}
+```
+
+### Polling
+Push notifications require full duplex connections (i.e., websocket or IPC).  Instead of subscribing, callers using HTTP send an `eth_newFilter` request:
+
+```json
+{"jsonrpc": "2.0", "id": 1, "method": "eth_newFilter", "params": ["returnData"]}
+```
+
+The Ethereum node responds with a filter ID:
+
+```json
+{"jsonrpc": "2.0", "id": 1, "result": "0x1"}
+```
+
+When a transaction is submitted, the Ethereum node computes the return data and pushes it to a queue, which is emptied when the caller polls using `eth_getFilterChanges`:
+
+```json
+{"jsonrpc": "2.0", "id": 2, "method": "eth_getFilterChanges", "params": ["0x1"]}
+```
+
+The node responds with an array of transaction hashes and their corresponding return data, in the order they were computed:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 2,
+  "result": [{
+    "transactionHash": "0x00000000000000000000000000000000000000000000000000000000deadbeef",
+    "returnData": "0x000000000000000000000000000000000000000000000000000000000000002a"
+  }]
+}
+```
+
+All transactions submitted by the client that were sealed _after_ the initial `eth_newFilter` request are included in this array.
+
+
+## Rationale
+[EIP 658](./eip-658.md) originally proposed adding return data to transaction receipts.  However, return data is not charged for (as it is not stored on the blockchain), so adding it to transaction receipts could result in DoS and spam opportunities.  Instead, a simple Boolean `status` field was added to transaction receipts.  This modified version of EIP 658 was included in the Byzantium hard fork.  While the `status` field is useful, applications often need the return data as well.
+
+The primary advantage of using the strategy outlined here is efficiency: no extra data needs to be stored on the blockchain, and minimal extra computational load is imposed on nodes.  Since light clients have the current state, they can compute and send return data notifications without contacting a server.  Although after-the-fact lookups of the return value would not be supported, this is consistent with the conventional use of return data, which are only accessible to the caller when the function returns, and are not stored for later use.
```

Comment:
Do light clients have the current state?  I thought they only maintain the block headers.

Reply:
You're correct.  Seems that I misunderstood how light clients work!  I have removed the incorrect statement about light clients.